onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -1,69 +1,87 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include "core/common/common.h"
 #include "core/optimizer/constant_folding.h"
 #include "core/graph/graph_utils.h"
 #include "core/optimizer/optimizer_execution_frame.h"
 #include "core/framework/op_kernel.h"
 #include "core/framework/ml_value.h"
 
+using namespace ::onnxruntime::common;
+
 namespace onnxruntime {
 
-Status ConstantFolding::Apply(Graph& graph, Node& node, bool& modified, bool& deleted) {
-  // Create execution frame for executing constant nodes.
-  OptimizerExecutionFrame::Info info({&node}, graph.GetAllInitializedTensors());
+Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  auto& order = graph_viewer.GetNodesInTopologicalOrder();
 
-  std::vector<int> fetch_mlvalue_idxs;
-  for (const auto* node_out : node.OutputDefs()) {
-    fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
-  }
+  for (NodeIndex i : order) {
+    auto* node = graph.GetNode(i);
+    if (!node) {
+      return Status(ONNXRUNTIME, INVALID_ARGUMENT);
+    }
 
-  OptimizerExecutionFrame frame(info, fetch_mlvalue_idxs);
+    // TODO Further investigate if this is needed here.
+    ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level));
 
-  auto* kernel = info.GetKernel(node.Index());
-  OpKernelContext op_kernel_context(&frame, kernel, ::onnxruntime::logging::LoggingManager::DefaultLogger());
+    // Check if constant folding can be applied on this node.
+    if (!graph_utils::IsSupportedProvider(*node, GetCompatibleExecutionProviders()) ||
+        excluded_op_types_.find(node->OpType()) != excluded_op_types_.end() ||
+        // constant folding is not currently supported in subgraphs.
```

Comment:
is it possible to check graph_level and return early in case of sub graph?

Reply:
Actually the problem are the control flow operators. We should be able to do constant folding in subgraphs today. But we cannot do in control flow operators that contain implicit inputs, as those are passed by the OpKernelContextInternal, which we should not use here.
So the comment is a bit misleading, will update it.